[Min Hein Aung] iP

MinHeinA opened the pull request:
### Duke

> "You can do anything, but not everything." – David Allen

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/MinHeinA/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```
public static void main(String[] args) {
    Duke duke = new Duke("data/duke.txt");
    duke.run();
}
```

Commit:
Add Gradle support

MinHeinA committed:
Complete Level 1. Greet, Echo, Exit

Implemented a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye

MinHeinA committed:
Complete Level 2. Add, List

Added the ability to store whatever text entered by the user and display them back to the user when requested.

MinHeinA committed:
Created Task.java

- Created Task.java and added necessary fields and its getters

MinHeinA committed:
Complete Level 3. Mark as Done

- Added the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

MinHeinA committed:
Added and Implements Deadline, Event and Todo classes

Added and Implemented Deadline, Event and Todo classes that inherits from Task class

MinHeinA committed:
Complete Level 4. ToDos, Events, Deadlines

Completed Level 4 by implementing the necessary logic in Duke.java file.
- Added: Events, Todo and Deadline tasks

MinHeinA committed:
Complete A-TextUiTesting

Completed A-TextUiTesting
-Used the input/output redirection technique to semi-automate the testing of Duke.

MinHeinA committed:
Created and Implemented Duke and Todo exception classes

Created and Implemented Duke and Todo exception classes

MinHeinA committed:
Complete Level 5. Handle Errors

The program now handles the following errors
- Unknown commands
- missing description for tasks (todo, event, deadline)
- missing index for mark/unmark
- blank user input

MinHeinA committed:
Complete Level 6. Delete

Added support for deleting tasks from the list.
The program now handles the following errors
- missing index for deletion of task
- index out of range for deletion of task

MinHeinA committed:
Complete A-Enums

Used enums to represent command keywords from user input

MinHeinA committed:
Complete Level 7. Save

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up. Data file location is [project_root]/data/duke.txt

MinHeinA committed:
Complete Level 8. Dates and Times

For Deadline and Event tasks, date data provided is now saved as LocalDateTime data. Duke now accepts date time input in dd-MM-yyyy HHmm and output the date data in MMM dd yyyy format.

MinHeinA committed:
Merge branch 'branch-Level-8'

MinHeinA committed:
Complete A-MoreOOP

Refactored code to extract out closely related code as class, classes Storage, TaskList and Ui has been refactored from Duke class

MinHeinA committed:
Compete stretch A-MoreOOP

- Extracted command classes with each command class having its own class that inherit from abstract command class.
- Rewrote main logic of app to work with changes.

MinHeinA committed:
Complete A-Packages

Organize the classes into suitable java packages.

MinHeinA committed:
Merge branch 'add-gradle-support'

MinHeinA committed:
Complete A-Gradle

Added gradle support for ip

MinHeinA committed:
Complete A-JUnit

Added A-JUnit for the classes Deadline, Event, Todo, Storage and TaskList.

MinHeinA committed:
Complete A-Jar

Configured shadow jar settings in build gradle

MinHeinA committed:
Complete A-JavaDoc

Added header comments to at least half of the non-private classes/methods.

MinHeinA committed:
Complete A-CodingStandard

Tweaked the code to comply with a coding standard

MinHeinA committed:
Complete Level-9

Gave users a way to find a task by searching for a keyword.

MinHeinA committed:
Merge branch 'branch-A-JavaDoc'

MinHeinA committed:
Merge branch 'branch-A-CodingStandard'

MinHeinA committed:
Merge branch 'branch-Level-9'

MinHeinA committed:
Complete A-CheckStyle

Configured CheckStyle in gradle and rectified style errors

MinHeinA committed:
Modify Gradle to add support for JavaFX

Modify Gradle to add support for JavaFX

lkwlkww commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+public class Duke {
+    public Ui ui;
+    public TaskList taskList;
+    public Storage storage;
```
I'm not sure if these variables are accessed by other classes within your implementation - if not, do you think the access modifier could be changed to `private`?

lkwlkww commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,133 @@
+                        String[] splitString = nextLine.split("\\|");
+
+                        Task task;
+                        switch (splitString[0]) {
```
Perhaps you could unindent the `case` lines here? According to the coding standard [guidelines](https://se-education.org/guides/conventions/java/index.html#variables):
> Note there is no indentation for case clauses.

lkwlkww commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.Scanner;
+
+public class Ui {
+    public static final String DIVIDER = "____________________________________________________________";
```
For my dividers, I used the `repeat` method of a String, which I thought would make it look cleaner. For example:
```java
"_".repeat(5) // Will become _____
```

My apologies for assuming you didn't know about this method. If you did, could you share why you chose not to do it that way?

tanhl2000 replied:
I actually like that you did not use the String.repeat() function as i think it improves readability of the code. However definitely understand that it is a QOL improvement to use the repeat() function and i also did not know about the repeat function before!

lkwlkww reviewed (commented):
Overall code quality seems really good!

MinHeinA committed:
Added FXML files and classes for JavaFX

Added the following FXML file
- MainWindow.fxml
- DialogBox.fxml

Added the following classes to prepare duke for JavaFX
- MainWindow.java
- Main.java
- DialogBox.java
- Launcher.java

MinHeinA committed:
Remove smiley emoticon

Removed smiley emoticon as it does not display display properly in JavaFX

MinHeinA committed:
Complete Level-10

- Modified command to return message instead of printing to terminal
- Added getResponse method that is called by MainWindow
- Upon bye command detect application will close in 3 seconds

MinHeinA committed:
Merge branch 'branch-A-CheckStyle'

MinHeinA committed:
Merge branch 'branch-Level-10'

tanhl2000 commented:
Overall code quality is good, coding standard is followed, other than the indentation of cases that was previously pointed out.

tanhl2000 commented:
I actually like that you did not use the String.repeat() function as i think it improves readability of the code. However definitely understand that it is a QOL improvement to use the repeat() function and i also did not know about the repeat function before

tanhl2000 commented:
for the DukeException however i believe it would be better if you could create individual exceptions for each exception case as it would make the exception handling structure better! right now a DukeException being thrown will not let us know anything unless we read the message that is also thrown, but i think the exception itself should give some information as well

tanhl2000 commented:
but overall really good work! :)

tanhl2000 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,21 @@
+ * A dedicated Exception class for Duke program.
+ *
+ */
+public class DukeException extends Exception {
```
for the DukeException however i believe it would be better if you could create individual exceptions for each exception case as it would make the exception handling structure better! right now a DukeException being thrown will not let us know anything unless we read the message that is also thrown, but i think the exception itself should give some information as well (sorry for the repeat comment I had not put it in the PR review section properly)

tanhl2000 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+package duke;
+
+import duke.command.Command;
```
Overall code quality is really good and coding standards are followed (other than the indentation fault in the case-switch section that was previously pointed out.

anuanas2007 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+
+    /**
+     * Starts the Duke program and requests for user input.
+     * If user types "bye", the program ends.
+     */
+    /*
+    public void run() {
+        ui.printGreetingMessage();
+        taskList = new TaskList(storage.loadTaskList());
+        String userInput;
+
+        while (true) {
+            userInput = ui.getUserInput();
+            try {
+                Command command = Parser.parse(userInput);
+                command.execute(taskList, storage, ui);
+                if (command.isExit()) {
+                    break;
+                }
+            } catch (DukeException exception) {
+                ui.printMessage(exception.toString() + "\n");
+            }
+        }
+    }
+     */
```
Perhaps you could decide whether to delete it out or not?

anuanas2007 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,46 @@
+            throw new DukeException("OOPS!!! The user input cannot be empty.");
+        }
+        command = userInput.split(" ")[0];
+        if (command.equals(Duke.Keyword.EXIT.getKeyword())) {
+            return new ExitCommand();
+        } else if (command.equals(Duke.Keyword.LIST.getKeyword())) {
+            return new ListCommand();
+        } else if (command.equals(Duke.Keyword.MARK.getKeyword()) || command.equals(Duke.Keyword.UNMARK.getKeyword())) {
+            return new MarkCommand(userInput, command.equals(Duke.Keyword.MARK.getKeyword()) ? true : false);
+        } else if (command.equals(Duke.Keyword.TODO.getKeyword())) {
+            return new AddCommand(userInput);
+        } else if (command.equals(Duke.Keyword.EVENT.getKeyword())) {
+            return new AddCommand(userInput);
+        } else if (command.equals(Duke.Keyword.DEADLINE.getKeyword())) {
+            return new AddCommand(userInput);
+        } else if (command.equals(Duke.Keyword.DELETE.getKeyword())) {
+            return new DeleteCommand(userInput);
+        } else if (command.equals(Duke.Keyword.FIND.getKeyword())) {
+            return new FindCommand(userInput);
+        }else {
+            throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
```
Perhaps using a switch statement might be an easier way to implement it?

anuanas2007 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+package duke;
+
```
Maybe you can consider changing your bot name?

anuanas2007 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,133 @@
+package duke;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
+
+/** Storage object used for file operations such as read and write of TaskList to text file. */
+public class Storage {
+
+    private Ui ui;
```
Storage class looks good!!

anuanas2007 commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,45 @@
+ * Represents the command to mark specific Tasks in TaskList as completed or not completed
+ * that inherits from Command.
+ */
+public class MarkCommand extends Command {
+    private final String userInput;
+    private final boolean isMark;
```
Maybe you consider changing the name of the class to better represent both marking and unmarking command.

anuanas2007 reviewed (commented):
Overall, your code looks good. Just a few suggestions you can consider!

clarence-chew commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+     */
+    public TaskList() {
+        this.taskList = new ArrayList<>();
+        this.ui = new Ui();
```
Coding standard recommends avoiding unnecessary use of `this` with fields.
https://se-education.org/guides/conventions/java/index.html#variables

clarence-chew commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+    public String getResponse(String userInput) {
+        try {
+            Command command = Parser.parse(userInput);
+            return command.execute(taskList, storage, ui);
```
I like how you made your Command class such that it is executed once at an obvious place.

clarence-chew commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,133 @@
+                        String nextLine = scanner.nextLine();
+                        String[] splitString = nextLine.split("\\|");
+
+                        Task task;
+                        switch (splitString[0]) {
+                        case "T":
+                            task = new Todo(splitString[2]);
+                            break;
+                        case "D": {
+                            task = new Deadline(splitString[2], splitString[3]);
+                            break;
+                        }
+                        case "E": {
+                            task = new Event(splitString[2], splitString[3]);
+                            break;
+                        }
+                        default:
+                            throw new DukeException("OOPS!!! Invalid task type found in file!");
+                        }
+
```
Perhaps some sort of Factory design pattern could be useful here?

MinHeinA committed:
Abstracted find task function and deleted text based UI class and functions

Abstracted find task function and deleted text based UI class and functions

MinHeinA committed:
Complete A Assertion

Added assertions

MinHeinA committed:
Merge pull request #1 from MinHeinA/branch-A-CodeQuality

Abstracted find task function and deleted text based UI class and fun…

MinHeinA committed:
Merge branch 'master' into branch-A-Assertions

MinHeinA committed:
Merge pull request #2 from MinHeinA/branch-A-Assertions

Complete A Assertion

MinHeinA committed:
Add priority to Duke tasks

- Add priority to Duke tasks
- Priority set during task creation, e.g. "Todo feed cat /priority high"
- If priority is not set during task creation e.g. "Todo feed cat", priority is automatically set as low
- Modify storage class to accommodate save and load of tasks with priority

MinHeinA committed:
Add priority command to change task priority

- Add priority command to change task priority
- Parser parse priority command

MinHeinA committed:
Update JUnit tests to accommodate tasks with priority

- Update JUnit tests to accommodate tasks with priority

MinHeinA committed:
Merge pull request #4 from MinHeinA/branch-C-Priority

Branch c priority

MinHeinA committed:
Add greeting message and change profile icons

Add greeting message and change profile icons

MinHeinA committed:
Add user guide README and move classes to folder

MinHeinA committed:
Merge branch 'branch-AddWelcomeMessage'

MinHeinA committed:
update user guide README

MinHeinA committed:
Update main README and add UI screenshot

MinHeinA committed:
Update main README.md

MinHeinA committed:
Add credit to code

MinHeinA committed:
Add application icon and title

MinHeinA committed:
Add more javadocs

MinHeinA committed:
Fix styling issues
